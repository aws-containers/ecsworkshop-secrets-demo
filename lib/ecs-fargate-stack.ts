import { App, Stack, StackProps, CfnOutput } from '@aws-cdk/core';
import { Vpc } from "@aws-cdk/aws-ec2";
import { Cluster, ContainerImage, Secret as ECSSecret } from "@aws-cdk/aws-ecs";
import { ApplicationLoadBalancedFargateService } from '@aws-cdk/aws-ecs-patterns';
import { Secret } from '@aws-cdk/aws-secretsmanager';

export interface ECSStackProps extends StackProps {
  vpc: Vpc
  dbSecretArn: string
}

export class ECSStack extends Stack {

  constructor(scope: App, id: string, props: ECSStackProps) {
    super(scope, id, props);

    const containerPort = this.node.tryGetContext("containerPort");
    const containerImage = this.node.tryGetContext("containerImage");
    const creds = Secret.fromSecretCompleteArn(this, 'postgresCreds', props.dbSecretArn);

    const cluster = new Cluster(this, 'Cluster', {
      vpc: props.vpc,
      clusterName: 'fargateClusterDemo'
    });

    const fargateService = new ApplicationLoadBalancedFargateService(this, "fargateService", {
      cluster,
      taskImageOptions: {
        image: ContainerImage.fromRegistry(containerImage),
        containerPort: containerPort,
        enableLogging: true,
        secrets: {
          POSTGRES_DATA: ECSSecret.fromSecretsManager(creds)
        }
      },
      desiredCount: 1,
      publicLoadBalancer: true,
      serviceName: 'fargateServiceDemo'
    });

    new CfnOutput(this, 'LoadBalancerDNS', { value: fargateService.loadBalancer.loadBalancerDnsName });
  }

}