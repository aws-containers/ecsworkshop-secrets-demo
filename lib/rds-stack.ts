import { App, StackProps, Stack, Duration, RemovalPolicy, CfnOutput } from "@aws-cdk/core";
import {
    DatabaseSecret, Credentials, ServerlessCluster, DatabaseClusterEngine, ParameterGroup, AuroraCapacityUnit
} from '@aws-cdk/aws-rds';
import { Vpc, Port, SubnetType } from '@aws-cdk/aws-ec2';
import { Secret, SecretRotation, SecretRotationApplication } from '@aws-cdk/aws-secretsmanager';

export interface RDSStackProps extends StackProps {
    vpc: Vpc
}

export class RDSStack extends Stack {

    readonly dbSecret: DatabaseSecret;
    readonly postgresRDSserverless: ServerlessCluster;

    constructor(scope: App, id: string, props: RDSStackProps) {
        super(scope, id, props);

        const dbUser = this.node.tryGetContext("dbUser");
        const dbName = this.node.tryGetContext("dbName");
        const dbPort = this.node.tryGetContext("dbPort") || 5432;

        this.dbSecret = new Secret(this, 'dbCredentialsSecret', {
            secretName: "ecsworkshop/test/todo-app/aurora-pg",
            generateSecretString: {
                secretStringTemplate: JSON.stringify({
                    username: dbUser,
                }),
                excludePunctuation: true,
                includeSpace: false,
                generateStringKey: 'password'
            }
        });

        this.postgresRDSserverless = new ServerlessCluster(this, 'postgresRdsServerless', {
            engine: DatabaseClusterEngine.AURORA_POSTGRESQL,
            parameterGroup: ParameterGroup.fromParameterGroupName(this, 'ParameterGroup', 'default.aurora-postgresql10'),
            vpc: props.vpc,
            enableDataApi: true,
            vpcSubnets: { subnetType: SubnetType.PRIVATE },
            credentials: Credentials.fromSecret(this.dbSecret, dbUser),
            scaling: {
                autoPause: Duration.minutes(10), // default is to pause after 5 minutes of idle time
                minCapacity: AuroraCapacityUnit.ACU_8, // default is 2 Aurora capacity units (ACUs)
                maxCapacity: AuroraCapacityUnit.ACU_32, // default is 16 Aurora capacity units (ACUs)
            },
            defaultDatabaseName: dbName,
            deletionProtection: false,
            removalPolicy: RemovalPolicy.DESTROY
        });

        this.postgresRDSserverless.connections.allowFromAnyIpv4(Port.tcp(dbPort));

        new SecretRotation(
            this,
            `ecsworkshop/test/todo-app/aurora-pg`,
            {
                secret: this.dbSecret,
                application: SecretRotationApplication.POSTGRES_ROTATION_SINGLE_USER,
                vpc: props.vpc,
                vpcSubnets: { subnetType: SubnetType.PRIVATE },
                target: this.postgresRDSserverless,
                automaticallyAfter: Duration.days(30),
            }
        );

        new CfnOutput(this, 'SecretName', { value: this.dbSecret.secretName });
    }
}