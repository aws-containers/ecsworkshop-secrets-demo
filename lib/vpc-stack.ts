import { App, Stack, StackProps, Construct } from '@aws-cdk/core';
import { Vpc, SubnetType } from '@aws-cdk/aws-ec2'

export interface VpcProps extends StackProps {
    maxAzs: number;
}

export class VPCStack extends Stack {
    readonly vpc: Vpc;

    constructor(scope: Construct, id: string, props: VpcProps) {
        super(scope, id, props);

        if (props.maxAzs !== undefined && props.maxAzs <= 1) {
            throw new Error('maxAzs must be at least 2.');
        }

        this.vpc = new Vpc(this, 'ecsWorkshopVPC', {
            cidr: "10.0.0.0/16",
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: 'public',
                    subnetType: SubnetType.PUBLIC,
                },
                {
                    cidrMask: 24,
                    name: 'private',
                    subnetType: SubnetType.PRIVATE,
                },
            ],
        });
    }
}