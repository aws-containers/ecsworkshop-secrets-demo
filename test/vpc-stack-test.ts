import { SynthUtils , expect as expectCDK, haveResource, countResources } from '@aws-cdk/assert';
import { App } from 'aws-cdk-lib';
import { VPCStack } from '../lib/vpc-stack'
import '@aws-cdk/assert/jest';

const iVpc = {
  maxAzs: 2
}

test('Fine-Grained Assertions', () => {
    const app = new App();
    // WHEN
    const stack = new VPCStack(app, 'VPCStack', iVpc);
    // THEN
    expectCDK(stack).to(haveResource('AWS::EC2::VPC', {
      EnableDnsHostnames: true,
      EnableDnsSupport: true,
      InstanceTenancy: "default",
    }));
});

test('Snapshot Tests', () => {
  const app = new App();
  const stack = new VPCStack(app, 'VPCStack', iVpc);
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Validation Tests - maxAzs = 0 is Error', () => {
  const app = new App();
  expect(() => {
    new VPCStack(app, 'VPCStack', {
      maxAzs: 0
    });
  }).toThrowError('maxAzs must be at least 2.');
});

test('Validation Tests - maxAzs = 1 is Error', () => {
  const app = new App();
  expect(() => {
    new VPCStack(app, 'VPCStack', {
      maxAzs: 1
    });
  }).toThrowError('maxAzs must be at least 2.');
});

test('countResources', () => {
  const app = new App();
  const stack = new VPCStack(app, 'VPCStack', iVpc);
  expectCDK(stack).to(countResources('AWS::EC2::Subnet', 4));
});