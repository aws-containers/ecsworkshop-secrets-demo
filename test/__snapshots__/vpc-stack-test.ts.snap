// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests 1`] = `
Object {
  "Resources": Object {
    "WorkshopVPCEB5EA86E": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "WorkshopVPCIGW461A4069": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "WorkshopVPCVPCGWE7DB3F39": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "WorkshopVPCIGW461A4069",
        },
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "WorkshopVPCprivateSubnet1DefaultRoute11A1F901": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "WorkshopVPCpublicSubnet1NATGateway66E0CE06",
        },
        "RouteTableId": Object {
          "Ref": "WorkshopVPCprivateSubnet1RouteTable5F1E4169",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "WorkshopVPCprivateSubnet1RouteTable5F1E4169": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/privateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "WorkshopVPCprivateSubnet1RouteTableAssociation3F94C6AC": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "WorkshopVPCprivateSubnet1RouteTable5F1E4169",
        },
        "SubnetId": Object {
          "Ref": "WorkshopVPCprivateSubnet1Subnet3A829D6F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "WorkshopVPCprivateSubnet1Subnet3A829D6F": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.2.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/privateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "WorkshopVPCprivateSubnet2DefaultRoute288AF3EA": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "WorkshopVPCpublicSubnet2NATGatewayEF44A90D",
        },
        "RouteTableId": Object {
          "Ref": "WorkshopVPCprivateSubnet2RouteTableFC058447",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "WorkshopVPCprivateSubnet2RouteTableAssociation2A9AEB9A": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "WorkshopVPCprivateSubnet2RouteTableFC058447",
        },
        "SubnetId": Object {
          "Ref": "WorkshopVPCprivateSubnet2Subnet786F2147",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "WorkshopVPCprivateSubnet2RouteTableFC058447": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/privateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "WorkshopVPCprivateSubnet2Subnet786F2147": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.3.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/privateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "WorkshopVPCpublicSubnet1DefaultRoute99487CA1": Object {
      "DependsOn": Array [
        "WorkshopVPCVPCGWE7DB3F39",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "WorkshopVPCIGW461A4069",
        },
        "RouteTableId": Object {
          "Ref": "WorkshopVPCpublicSubnet1RouteTableA683EC6E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "WorkshopVPCpublicSubnet1EIP7158FBF0": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "WorkshopVPCpublicSubnet1NATGateway66E0CE06": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "WorkshopVPCpublicSubnet1EIP7158FBF0",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "WorkshopVPCpublicSubnet1Subnet121CA209",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "WorkshopVPCpublicSubnet1RouteTableA683EC6E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "WorkshopVPCpublicSubnet1RouteTableAssociation5BCDAD57": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "WorkshopVPCpublicSubnet1RouteTableA683EC6E",
        },
        "SubnetId": Object {
          "Ref": "WorkshopVPCpublicSubnet1Subnet121CA209",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "WorkshopVPCpublicSubnet1Subnet121CA209": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "WorkshopVPCpublicSubnet2DefaultRoute4C6BE8AC": Object {
      "DependsOn": Array [
        "WorkshopVPCVPCGWE7DB3F39",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "WorkshopVPCIGW461A4069",
        },
        "RouteTableId": Object {
          "Ref": "WorkshopVPCpublicSubnet2RouteTableCD447869",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "WorkshopVPCpublicSubnet2EIP30EC892B": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "WorkshopVPCpublicSubnet2NATGatewayEF44A90D": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "WorkshopVPCpublicSubnet2EIP30EC892B",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "WorkshopVPCpublicSubnet2Subnet91B14F62",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "WorkshopVPCpublicSubnet2RouteTableAssociation41EF4ED8": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "WorkshopVPCpublicSubnet2RouteTableCD447869",
        },
        "SubnetId": Object {
          "Ref": "WorkshopVPCpublicSubnet2Subnet91B14F62",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "WorkshopVPCpublicSubnet2RouteTableCD447869": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "WorkshopVPCpublicSubnet2Subnet91B14F62": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "VPCStack/WorkshopVPC/publicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "WorkshopVPCEB5EA86E",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
